File: services/csharp/graphql/src/application/service/ProductService.cs
using src.application.interfaces;
using src.domain.entities;
using src.domain.interfaces;
using src.graphql.inputs;
using src.mappings;

namespace src.application.service;

public class ProductService : IProductService
{
    private readonly IProductRepository _productRepository;

    public ProductService(IProductRepository productRepository)
    {
        _productRepository = productRepository;
    }

    public Product GetProductById(int id)
    {
        return _productRepository.GetById(id)
            ?? new Product
            {
                Id = id,
                Name = $"Product {id}",
                Description = "Product not found in sample list",
                Price = 0,
                StockQuantity = 0,
                CreatedAt = DateTime.UtcNow
            };
    }

    public IEnumerable<Product> GetAllProducts()
    {
        return _productRepository.GetAll();
    }

    public Product InsertProduct(CreateProductInput request)
    {
        var product = request.ToEntity();
        return _productRepository.Add(product);
    }

    public Product UpdateProduct(int id, UpdateProductInput request)
    {
        var product = _productRepository.GetById(id);

        if (product is null)
        {
            product = request.ToEntityFromUpdate(id);
            return _productRepository.Add(product);
        }

        request.UpdateEntity(product);
        return _productRepository.Update(product);
    }

    public bool DeleteProduct(int id)
    {
        return _productRepository.Delete(id);
    }

    public bool DeleteProduct()
    {
        return _productRepository.DeleteAll();
    }
}

File: services/go/graphql/src/application/service/product_service.go
package service

import (
	appinterfaces "graphql/src/application/interfaces"
	"graphql/src/domain/entities"
	domaininterfaces "graphql/src/domain/interfaces"
)

type ProductService struct {
	repository domaininterfaces.ProductRepository
}

func NewProductService(repository domaininterfaces.ProductRepository) *ProductService {
	return &ProductService{repository: repository}
}

func (s *ProductService) GetByID(id int) (*entities.Product, error) {
	return s.repository.GetByID(id)
}

func (s *ProductService) GetAll() ([]entities.Product, error) {
	return s.repository.GetAll()
}

func (s *ProductService) Create(input entities.CreateInput) (*entities.Product, error) {
	return s.repository.Create(input)
}

func (s *ProductService) Update(id int, input entities.UpdateInput) (*entities.Product, error) {
	return s.repository.Update(id, input)
}

func (s *ProductService) Delete(id int) (bool, error) {
	return s.repository.Delete(id)
}

func (s *ProductService) DeleteAll() (bool, error) {
	return s.repository.DeleteAll()
}

var _ appinterfaces.ProductService = (*ProductService)(nil)
